File: src_tf/demo_cpp_tf/src/static_tf_broadcaster.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp> //提供消息接口
#include <tf2/LinearMath/Quaternion.h> //四元数类
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp> //四元数与消息的转换
#include <tf2_ros/static_transform_broadcaster.h> //静态坐标容器


class StaticTFBroadcasterNode : public rclcpp::Node
{
private:
    std::shared_ptr<tf2_ros::StaticTransformBroadcaster> static_tf_broadcaster_;

public:
    StaticTFBroadcasterNode() : Node("static_tf_broadcaster")
    {
        this->static_tf_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);//这里输入节点this作为参数的作用是：函数会自动创建一个发布者取发布一个话题tf_static
        publish_tf();
    }

    void publish_tf()
    {
        geometry_msgs::msg::TransformStamped transform;

        transform.header.stamp = this->get_clock()->now();
        transform.header.frame_id = "map";
        transform.child_frame_id = "target_point";

        transform.transform.translation.x = 5.0;
        transform.transform.translation.y = 3.0;
        transform.transform.translation.z = 0.0;

        tf2::Quaternion q;
        q.setRPY(0.0, 0.0, 60*M_PI/180.0); // Roll, Pitch, Yaw
        transform.transform.rotation = tf2::toMsg(q);

        this->static_tf_broadcaster_->sendTransform(transform);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<StaticTFBroadcasterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src_tf/demo_python_tf/demo_python_tf/static_tf_broadcaster.py
import rclpy
from rclpy.node import Node
from tf2_ros import StaticTransformBroadcaster #静态坐标发布器
from geometry_msgs.msg import TransformStamped #消息接口
from tf_transformations import quaternion_from_euler #欧拉角转四元数
import math 

class StaticTFBroadCaster(Node):
    def __init__(self):
        super().__init__('static_tf_broadcaster')
        self.static_broadcaster_ = StaticTransformBroadcaster(self)
        self.publish_static_tf()

    #发布静态tf
    def publish_static_tf(self):
        #从base_link到camera_link的坐标关系
        transform = TransformStamped()

        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = 'base_link' #父坐标系名称
        transform.child_frame_id = 'camera_link' #子坐标系名称

        transform.transform.translation.x = 0.5
        transform.transform.translation.y = 0.3
        transform.transform.translation.z = 0.6

        #欧拉角转四元数 返回值为元组 xyzw
        q = quaternion_from_euler(math.radians(180), 0, 0) #math用于角度转弧度
        transform.transform.rotation.x = q[0]
        transform.transform.rotation.y = q[1]
        transform.transform.rotation.z = q[2]
        transform.transform.rotation.w = q[3]

        #发布静态坐标变换
        self.static_broadcaster_.sendTransform(transform)
        self.get_logger().info(f'{transform}')
def main():
    rclpy.init()
    node = StaticTFBroadCaster()
    rclpy.spin(node)
    rclpy.shutdown()
